pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,114 @@
+package nstemplateset
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	projectv1 "github.com/openshift/api/project/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_nstemplateset")
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNSTemplateSet{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nstemplateset-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.NSTemplateSet{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
```

Comment:
I think we will need to watch some objects as secondary resources here... (Namespaces, Projects?). We will need it for tracking the status. We should set the status as provisioned when everything is ready.

We will also need to set the UserAccount status to Ready when NSTemplateSet is ready.

Basically, our end-goal is to set UserAccount to Ready when all the underling resources are provisioned and ready to be used. It's fine to handle that status sync across all these resources in a separate PR though because it can be a bit tricky to implement. But if we do it in a separate PR then please create a JIRA for that so we don't forget it.

Reply:
@alexeykazakov thanks for input, I will consider this when start working on status handling.

Reply:
@alexeykazakov done here https://github.com/codeready-toolchain/member-operator/pull/53/commits/78e1e29931e679c690857910c384497288ad45e6

Namespace added as secondary resource.
Labels added for namespace.
One namespace creation in one reconcile.